adapters/handlers/rest/handlers_authz_add_permission_test.go

```diff
@@ -0,0 +1,270 @@
+				Body: authz.AddPermissionsBody{
+					Name:        String("someName"),
+					Permissions: []*models.Permission{},
+				},
+			},
+			principal:     &models.Principal{Username: "user1"},
+			expectedError: "role has to have at least 1 permission",
+		},
+		{
+			name: "invalid action",
+			params: authz.AddPermissionsParams{
+				Body: authz.AddPermissionsBody{
+					Name: String("someName"),
+					Permissions: []*models.Permission{
+						{
+							Action: String("manage_somethingelse"),
+						},
+					},
+				},
+			},
+			principal:     &models.Principal{Username: "user1"},
+			expectedError: "invalid permission",
+		},
+		{
+			name: "update builtin role",
+			params: authz.AddPermissionsParams{
+				Body: authz.AddPermissionsBody{
+					Name: &authorization.BuiltInRoles[0],
+					Permissions: []*models.Permission{
+						{
+							Action: String("manage_roles"),
+						},
+					},
+				},
+			},
+			principal:     &models.Principal{Username: "user1"},
+			expectedError: "you can not update builtin role",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			authorizer := mocks.NewAuthorizer(t)
+			controller := mocks.NewController(t)
+			logger, _ := test.NewNullLogger()
+
+			authorizer.On("Authorize", tt.principal, authorization.UPDATE, authorization.Roles(*tt.params.Body.Name)[0]).Return(nil)
+
+			h := &authZHandlers{
+				authorizer: authorizer,
+				controller: controller,
+				logger:     logger,
+			}
+			res := h.addPermissions(tt.params, tt.principal)
+			parsed, ok := res.(*authz.AddPermissionsBadRequest)
+			assert.True(t, ok)
+
+			if tt.expectedError != "" {
+				assert.Contains(t, parsed.Payload.Error[0].Message, tt.expectedError)
+			}
+		})
+	}
+}
+
+func TestAddPermissionsForbidden(t *testing.T) {
+	type testCase struct {
+		name          string
+		params        authz.AddPermissionsParams
+		principal     *models.Principal
+		authorizeErr  error
+		expectedError string
+	}
+
+	tests := []testCase{
+		{
+			name: "update some role",
+			params: authz.AddPermissionsParams{
+				Body: authz.AddPermissionsBody{
+					Name: String("someRole"),
+					Permissions: []*models.Permission{
+						{
+							Action: String("manage_roles"),
+						},
+					},
+				},
+			},
+			principal:     &models.Principal{Username: "user1"},
+			authorizeErr:  fmt.Errorf("some error from authZ"),
+			expectedError: "some error from authZ",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			authorizer := mocks.NewAuthorizer(t)
+			controller := mocks.NewController(t)
+			logger, _ := test.NewNullLogger()
+
+			authorizer.On("Authorize", tt.principal, authorization.UPDATE, authorization.Roles(*tt.params.Body.Name)[0]).Return(tt.authorizeErr)
+
+			h := &authZHandlers{
+				authorizer: authorizer,
+				controller: controller,
+				logger:     logger,
+			}
+			res := h.addPermissions(tt.params, tt.principal)
+			parsed, ok := res.(*authz.AddPermissionsForbidden)
```

Comment:
does this mean that the authorizer/casbin says that the user is not allowed to manage roles?